bn-apps/memberships-management/membership-management-demo/src/main/kotlin/com/r3/businessnetworks/memberships/demo/flows/Flows.kt

```diff
@@ -0,0 +1,124 @@
+import net.corda.core.transactions.SignedTransaction
+import net.corda.core.transactions.TransactionBuilder
+import net.corda.core.utilities.unwrap
+
+@CordaSerializable
+class RequestMembershipForTransaction (val bno : Party)
+
+class IssueAssetFlow (val bno : Party) : FlowLogic<SignedTransaction>() {
+    @Suspendable
+    override fun call() : SignedTransaction {
+        // for the sake of demo we are just taking the first available notary
+        val notary = serviceHub.networkMapCache.notaryIdentities.single()
+
+        // query the vault with bno
+        val memberships= subFlow(GetMembershipsFlow(bno))
+        val ourMembership = memberships[ourIdentity]!!
+
+        val builder = TransactionBuilder(notary)
+                .addOutputState(SampleState(ourIdentity), SampleContract.CONTRACT_NAME)
+                .addCommand(SampleContract.Commands.Issue(), ourIdentity.owningKey)
+                .addReferenceState(ReferencedStateAndRef(ourMembership))
+
+        builder.verify(serviceHub)
+        val stx = serviceHub.signInitialTransaction(builder)
+        return subFlow(FinalityFlow(stx, listOf()))
+    }
+}
+
+@InitiatingFlow
+class TransferAssetFlow(val sampleState : StateAndRef<SampleState>,
+                              val bno : Party,
+                              val partyToTransferTo : Party) : FlowLogic<SignedTransaction>() {
+    @Suspendable
+    override fun call() : SignedTransaction {
+        if (ourIdentity != sampleState.state.data.owner)
+            throw FlowException("Only owner of the state can transfer it")
+        if (ourIdentity == partyToTransferTo)
+            throw FlowException("sender and recipient should be different parties")
+        // for the sake of demo we are just taking the first available notary
+        val notary = serviceHub.networkMapCache.notaryIdentities.single()
+
+        // get the membership from the node's membership cache
+        val membershipsCacheHolderService = serviceHub.cordaService(MembershipsCacheHolder::class.java)
+        val ourMembership = membershipsCacheHolderService.cache.getMembership(bno, ourIdentity)
+
+        if (ourMembership == null) {
+           throw FlowException("Invalid membership: ourMembership is null")
+        }
+
+        val builder = TransactionBuilder(notary)
+                .addReferenceState(ReferencedStateAndRef(ourMembership))
+                .addInputState(sampleState)
+                .addOutputState(SampleState(partyToTransferTo))
+                .addCommand(SampleContract.Commands.Transfer(), ourIdentity.owningKey, partyToTransferTo.owningKey)
+
+        // create a session
+        val counterPartySession = initiateFlow(partyToTransferTo)
+        // requesting other party to send us their billing states
+        counterPartySession.send(RequestMembershipForTransaction(bno))
+
+
+        val counterPartyMembershipState = subFlow(ReceiveStateAndRefFlow<MembershipState<Any>>(counterPartySession)).single()
```

Comment:
Why not to get this state directly from the memberships cache?

Reply:
See above. That's what I tried to do initially. But if you do that the transaction resolution fails because you do not have the chain for the counterparty's membership.